include/slang/binding/OperatorExpressions.h

```diff
@@ -248,20 +248,35 @@ class ReplicationExpression : public Expression {
 
 struct StreamingConcatenationExpressionSyntax;
 
+#define RANGE(x) x(Simple) x(IndexedUp) x(IndexedDown) x(Bit)
+ENUM(WithRangeKind, RANGE); // RangeSelectionKind + Bit
+#undef RANGE
+
 /// Represents a streaming concatenation
 class StreamingConcatenationExpression : public Expression {
 public:
     /// The size of the blocks to slice and reorder: if 0, this is a left-to-right
     /// concatenation. Otherwise, it's a right-to-left concatenation.
     const size_t sliceSize;
 
+    struct WithExpression {
+        WithRangeKind kind;
+        Expression* left;
+        Expression* right;
+        optional<int32_t> width; // constant width
+    };
+    struct StreamExpression {
+        Expression* operand;
+        WithExpression* with;
```

Comment:
These objects aren't very big, perhaps it would be better to make them optional instead of heap allocating?

Reply:
Yes, I replace it with
```c++
        optional<WithExpression> with;
```

Reply:
However, I am not sure whether it makes a big difference. When "with" is a pointer, I use the bump allocator
```c++
 with = compilation.emplace<WithExpression>(WithExpression{ kind, left, right, width });
```
If "with" is optional, it is heap allocated together in
```c++
    return *compilation.emplace<StreamingConcatenationExpression>(
        compilation.getVoidType(), sliceSize, buffer.ccopy(compilation), syntax.sourceRange());
```
Eventually both methods are heap allocated by the bump allocator. The difference is only between allocated separately or allocated together.

Reply:
I reversed the change for now. I have not fully understood the benefit of the change, because both methods use bump allocator on heap. I will make the change if you confirm there is a benefit doing so.

Reply:
It's slightly more efficient to not do the extra pointer indirections, but yeah, the bump allocator definitely makes the heap allocations much less costly. It's fine with me to leave it this way.